hardware_interface/test/fake_components/test_generic_system.cpp

```diff
@@ -1027,3 +1100,177 @@ TEST_F(TestGenericSystem, generic_system_2dof_functionality_with_offset_custom_i
   EXPECT_EQ(
     states_map["GenericSystem2dof"].label(), hardware_interface::lifecycle_state_names::INACTIVE);
 }
+
+TEST_F(TestGenericSystem, valid_urdf_ros2_control_system_robot_with_gpio_)
+{
+  auto urdf = ros2_control_test_assets::urdf_head +
+              valid_urdf_ros2_control_system_robot_with_gpio_ + ros2_control_test_assets::urdf_tail;
+  hardware_interface::ResourceManager rm(urdf);
+
+  // check is hardware is configured
+  std::unordered_map<std::string, rclcpp_lifecycle::State> states_map;
+  states_map = rm.get_components_states();
+  EXPECT_EQ(
+    states_map["GenericSystem2dof"].label(), hardware_interface::lifecycle_state_names::INACTIVE);
+
+  // Check initial values
+  hardware_interface::LoanedStateInterface gpio1_a_o1_s =
+    rm.claim_state_interface("flange_analog_IOs/analog_output1");
+  hardware_interface::LoanedStateInterface gpio1_a_i1_s =
+    rm.claim_state_interface("flange_analog_IOs/analog_input1");
+  hardware_interface::LoanedStateInterface gpio1_a_o2_s =
+    rm.claim_state_interface("flange_analog_IOs/analog_input2");
+  hardware_interface::LoanedStateInterface gpio2_vac_s =
+    rm.claim_state_interface("flange_vacuum/vacuum");
+  hardware_interface::LoanedCommandInterface gpio1_a_o1_c =
+    rm.claim_command_interface("flange_analog_IOs/analog_output1");
+  hardware_interface::LoanedCommandInterface gpio2_vac_c =
+    rm.claim_command_interface("flange_vacuum/vacuum");
+
+  // State interfaces without initial value are set to 0
+  ASSERT_EQ(0.0, gpio1_a_o1_s.get_value());
+  ASSERT_EQ(0.0, gpio2_vac_s.get_value());
+  ASSERT_TRUE(std::isnan(gpio1_a_o1_c.get_value()));
+  ASSERT_TRUE(std::isnan(gpio2_vac_c.get_value()));
+
+  // set some new values in commands
+  gpio1_a_o1_c.set_value(0.111);
+  gpio2_vac_c.set_value(0.222);
+
+  // State values should not be changed
+  ASSERT_EQ(0.0, gpio1_a_o1_s.get_value());
+  ASSERT_EQ(0.0, gpio2_vac_s.get_value());
```

Comment:
Shouldn't this state have value "NaN" because it does not have initial value. Why did you decide to set them to 0?

Reply:
This is being set at line https://github.com/ros-controls/ros2_control/blob/41f3cce1a9c3c6ec23ea51a19c4a58d65e0ad70a/hardware_interface/src/fake_components/generic_system.cpp#L210

Do we keep it as NaN?

Reply:
yes please

Reply:
If it make sense

Reply:
Done.